metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -0,0 +1,456 @@
+            fenced = prevRegistration.fenced();
+        }
+        // Update broker registrations.
+        brokerRegistrations.put(brokerId, new BrokerRegistration(brokerId,
+            record.brokerEpoch(), record.incarnationId(), listeners, features,
+            Optional.ofNullable(record.rack()), fenced));
+
+        if (prevRegistration == null) {
+            log.info("Registered new broker: {}", record);
+        } else if (prevRegistration.incarnationId().equals(record.incarnationId())) {
+            log.info("Re-registered broker incarnation: {}", record);
+        } else {
+            log.info("Re-registered broker id {}: {}", brokerId, record);
+        }
+    }
+
+    public void replay(UnregisterBrokerRecord record) {
+        int brokerId = record.brokerId();
+        BrokerRegistration registration = brokerRegistrations.get(brokerId);
+        if (registration == null) {
+            throw new RuntimeException(String.format("Unable to replay %s: no broker " +
+                "registration found for that id", record.toString()));
+        } else if (registration.epoch() !=  record.brokerEpoch()) {
+            throw new RuntimeException(String.format("Unable to replay %s: no broker " +
+                "registration with that epoch found", record.toString()));
+        } else {
+            brokerRegistrations.remove(brokerId);
+            log.info("Unregistered broker: {}", record);
+        }
+    }
+
+    public void replay(FenceBrokerRecord record) {
+        int brokerId = record.id();
+        BrokerRegistration registration = brokerRegistrations.get(brokerId);
+        if (registration == null) {
+            throw new RuntimeException(String.format("Unable to replay %s: no broker " +
+                "registration found for that id", record.toString()));
+        } else if (registration.epoch() !=  record.epoch()) {
+            throw new RuntimeException(String.format("Unable to replay %s: no broker " +
+                "registration with that epoch found", record.toString()));
+        } else {
+            brokerRegistrations.put(brokerId, registration.cloneWithFencing(true));
+            log.info("Fenced broker: {}", record);
+        }
+    }
+
+    public void replay(UnfenceBrokerRecord record) {
+        int brokerId = record.id();
+        BrokerRegistration registration = brokerRegistrations.get(brokerId);
+        if (registration == null) {
+            throw new RuntimeException(String.format("Unable to replay %s: no broker " +
+                "registration found for that id", record.toString()));
+        } else if (registration.epoch() !=  record.epoch()) {
+            throw new RuntimeException(String.format("Unable to replay %s: no broker " +
+                "registration with that epoch found", record.toString()));
+        } else {
+            brokerRegistrations.put(brokerId, registration.cloneWithFencing(false));
```

Comment:
Good point. The appropriate place to handle this will be in the broker heartbeat handling code, since that is where the active controller unfences brokers.